src/Exceptions/ExceptionWithContext.php

```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace Cascade\Exceptions;
+
+/**
+ * A custom exception class which provides a context argument for later reference.
+ */
+class ExceptionWithContext extends \Exception
```

Comment:
Can you say more about the odd feeling? I think it doesn't bother me (though catching a subclass, or catching \Exception and `instanceof` checking is probably more likely in a lot of cases).

I wouldn't make it abstract on the argument that `\Exception` isn't abstract either, and I don't think there's a real reason to prevent people from throwing an instance of it.

The interface is an option, but I think it also suggests a trait to make it easy to embed contexts in exceptions and I'm not sure I prefer two structural objects instead of one to implement this idea.